File: app/src/main/java/tw/com/chainsea/chat/messagekit/main/viewholder/VideoMessageView.java
//package tw.com.chainsea.chat.messagekit.main.viewholder;
//
//import android.os.Environment;
//import android.view.LayoutInflater;
//import android.view.View;
//import android.view.ViewGroup;
//import android.webkit.URLUtil;
//import android.widget.MediaController;
//
//import androidx.annotation.NonNull;
//import androidx.core.content.res.ResourcesCompat;
//import androidx.viewbinding.ViewBinding;
//
//import com.bumptech.glide.Glide;
//import com.bumptech.glide.load.engine.DiskCacheStrategy;
//import com.bumptech.glide.request.RequestOptions;
//import com.google.common.base.Strings;
//import com.google.common.primitives.Ints;
//import com.luck.picture.lib.utils.ToastUtils;
//
//import java.io.File;
//
//import tw.com.aile.sdk.bean.message.MessageEntity;
//import tw.com.chainsea.android.common.log.CELog;
//import tw.com.chainsea.android.common.system.ThreadExecutorHelper;
//import tw.com.chainsea.android.common.ui.UiHelper;
//import tw.com.chainsea.android.common.video.VideoHelper;
//import tw.com.chainsea.ce.sdk.bean.msg.MessageStatus;
//import tw.com.chainsea.ce.sdk.bean.msg.content.VideoContent;
//import tw.com.chainsea.ce.sdk.database.sp.TokenPref;
//import tw.com.chainsea.chat.R;
//import tw.com.chainsea.chat.databinding.MsgkitVideoBinding;
//import tw.com.chainsea.chat.messagekit.main.viewholder.base.MessageBubbleView;
//import tw.com.chainsea.chat.util.DownloadUtil;
//
///**
// * current by evan on 2020-01-08
// */
//public class VideoMessageView extends MessageBubbleView<VideoContent> {
//    final static int SMALL_PIC_SIZE = 360;
//    static int MARGIN = 0;
//    private final String downloadDir = Environment.getExternalStorageDirectory() + "/download/";
//    private DownloadingStatus newestStatus = DownloadingStatus.UNDEF;
//    private enum DownloadingStatus {
//        UNDEF, SUCCESS, CANCELED, FAILURE
//    }
//
//    private MsgkitVideoBinding binding;
//    public VideoMessageView(@NonNull ViewBinding binding) {
//        super(binding);
//        this.binding = MsgkitVideoBinding.inflate(LayoutInflater.from(itemView.getContext()), null, false);
//        getView(this.binding.getRoot());
//        MARGIN = UiHelper.dip2px(binding.getRoot().getContext(), 8);
//    }
//
//    @Override
//    protected int getContentResId() {
//        return R.layout.msgkit_video;
//    }
//
//    @Override
//    protected int rightBackground() {
//        return R.drawable.msgkit_trans_bg;
//    }
//
//    @Override
//    protected int leftBackground() {
//        return R.drawable.msgkit_trans_bg;
//    }
//
//    @Override
//    protected View getChildView() {
//        return binding.getRoot();
//    }
//
//    @Override
//    protected boolean showName() {
//        return isRightMessage() ? false : true;
//    }
//
//    @Override
//    protected void bindContentView(VideoContent videoContent) {
//        String thumbnailUrl = "";
//        int thumbnailWidth = 300;
//        int thumbnailHeight = 600;
//        ViewGroup.LayoutParams params = binding.occupationCL.getLayoutParams();
//        int[] zoomSize;
//        if (videoContent.getThumbnailHeight() == 0 || videoContent.getThumbnailWidth() == 0) {
//            zoomSize = VideoHelper.size(videoContent.getUrl());
//        } else if (videoContent.getWidth() != 0 || videoContent.getHeight() != 0) {
//            zoomSize = zoomImage(videoContent.getWidth(), videoContent.getHeight());
//        } else {
//            zoomSize = zoomImage(videoContent.getThumbnailWidth(), videoContent.getThumbnailHeight());
//        }
//
//        if(zoomSize[0] > 800){
//            params.width = SMALL_PIC_SIZE;
//            params.height = SMALL_PIC_SIZE;
//        }else {
//            params.width = zoomSize[0];
//            params.height = zoomSize[1];
//        }
//        // 如果沒高度寬度 給預設
//        if (params.width <= 0) params.width = thumbnailWidth;
//        if (params.height <= 0) params.height = thumbnailHeight;
//
//        binding.occupationCL.setLayoutParams(params);
//
//        MediaController mediaController = new MediaController(binding.getRoot().getContext());
//        mediaController.setVisibility(View.GONE);
//        String localPath = videoContent.getAndroid_local_path();
//        String videoUrl = videoContent.getUrl();
//
//        binding.occupationCL.setBackground(ResourcesCompat.getDrawable(binding.getRoot().getContext().getResources(), R.drawable.file_msg_down_bg, null));
////        binding.reviewVV.setVisibility(View.GONE);
//        binding.thumbnailRIV.setVisibility(View.VISIBLE);
//        binding.playIV.setVisibility(View.VISIBLE);
//
//        String progress = videoContent.getProgress();
//        MessageStatus status = getMessage().getStatus();
//        switch (status) {
//            case READ:
//            case RECEIVED:
//            case SUCCESS:
//                if (!Strings.isNullOrEmpty(progress)) {
//                    int pgs = Ints.tryParse(progress);
//                    if (pgs >= 100) {
//                        binding.progressBar.setVisibility(View.GONE);
//                        binding.playIV.setVisibility(View.VISIBLE);
//                    } else {
//                        binding.playIV.setVisibility(View.GONE);
//                        binding.progressBar.setProgress(pgs);
//                    }
//                }
//                break;
//            case SENDING:
//                if (!Strings.isNullOrEmpty(progress)) {
//                    int pgs = Ints.tryParse(progress);
//                    if (pgs >= 100) {
//                        binding.progressBar.setProgress(0);
//                        binding.progressBar.setVisibility(View.GONE);
//                        binding.playIV.setVisibility(View.VISIBLE);
//                    } else {
//                        binding.playIV.setVisibility(View.GONE);
//                        binding.progressBar.setProgress(pgs);
//                    }
//                }
//                break;
//            case FAILED:
//            case ERROR:
//            default:
//                break;
//        }
//
//
//        if (!Strings.isNullOrEmpty(localPath) && new File(localPath).exists()) {
//            binding.playIV.setImageResource(R.drawable.play);
//            Glide.with(binding.thumbnailRIV)
//                    .load(localPath)
//                    .apply(new RequestOptions()
//                            .frame(1000)
//                            .override(SMALL_PIC_SIZE)
//                            .diskCacheStrategy(DiskCacheStrategy.ALL)
//                            .fitCenter())
//                    .into(binding.thumbnailRIV);
//        } else if (URLUtil.isValidUrl(videoUrl) || !videoUrl.endsWith(".mp4")) {
//
//            if(getMessage().getContent()!=null && !getMessage().getContent().isEmpty()) {
//                try {
//                    thumbnailUrl = videoContent.getThumbnailUrl();
//                    if(videoContent.getThumbnailWidth() > 0 || videoContent.getThumbnailHeight() > 0) {
//                        thumbnailWidth = videoContent.getThumbnailWidth();
//                        thumbnailHeight = videoContent.getThumbnailHeight();
//                    }
//                }catch (Exception e) {
//                    CELog.e("VideoMessageView json parse error="+e.getMessage());
//                }
//            }
//
//            if(thumbnailUrl.isEmpty()){
//                Glide.with(binding.thumbnailRIV)
//                        .load(localPath)
//                        .apply(new RequestOptions()
//                                .frame(1000)
//                                .override(SMALL_PIC_SIZE)
//                                .diskCacheStrategy(DiskCacheStrategy.ALL)
//                                .fitCenter())
//                        .into(binding.thumbnailRIV);
//            }else{
//                Glide.with(binding.thumbnailRIV)
//                        .load(thumbnailUrl)
//                        .apply(new RequestOptions()
//                                .frame(1000)
//                                .override(thumbnailWidth, thumbnailHeight)
//                                .diskCacheStrategy(DiskCacheStrategy.ALL)
//                                .fitCenter())
//                        .into(binding.thumbnailRIV);
//            }
//            //判斷是否已下載
//            String path = downloadDir + getMessage().getSendTime() + "_" + videoContent.getName();
//            binding.playIV.setImageResource( new File(path).exists() ? R.drawable.play : R.drawable.ic_video_download);
//
//        } else {
//            binding.playIV.setVisibility(View.GONE);
//            Glide.with(binding.thumbnailRIV)
//                    .load(R.drawable.image_load_error)
//                    .apply(new RequestOptions().override(SMALL_PIC_SIZE))
//                    .into(binding.thumbnailRIV);
//        }
//    }
//
//    public static int[] zoomImage(int inWidth, int inHeight) {
//        int outWidth;
//        int outHeight;
//        if (inWidth > inHeight) {
//            outWidth = SMALL_PIC_SIZE;
//            outHeight = (inHeight * SMALL_PIC_SIZE) / inWidth;
//        } else {
//            outHeight = SMALL_PIC_SIZE;
//            outWidth = (inWidth * SMALL_PIC_SIZE) / inHeight;
//        }
//        int[] sizes = new int[2];
//        sizes[0] = outWidth;
//        sizes[1] = outHeight;
//        return sizes;
//    }
//
//
//    public void updateProgress(int progress) {
//        if (binding != null && binding.progressBar != null) {
//            ThreadExecutorHelper.getMainThreadExecutor().execute(() -> {
//                binding.progressBar.setProgress(progress);
//                if (binding.progressBar.getProgress() == 100) {
//                    binding.progressBar.setVisibility(View.GONE);
//                    binding.playIV.setVisibility(View.VISIBLE);
//                } else {
//                    binding.progressBar.setVisibility(View.VISIBLE);
//                    binding.playIV.setVisibility(View.GONE);
//                }
//            });
//        }
//    }
//
//    @Override
//    public void onClick(View v, MessageEntity message) {
//        super.onClick(v, message);
//
//        VideoContent format = (VideoContent) message.content();
//        String downloadPath = downloadDir + message.getSendTime() + "_" + format.getName();
//        File downloadFile = new File(downloadPath);
//        if ((!Strings.isNullOrEmpty(format.getAndroid_local_path()) && new File(format.getAndroid_local_path()).exists()) || downloadFile.exists()) {
//            if (this.onMessageControlEventListener != null) {
//                this.onMessageControlEventListener.onVideoClick(getMessage());
//            }
//            binding.progressBar.setOnFileClickListener(null);
//        } else {
//            binding.progressBar.setOnFileClickListener(() -> {
//                if (!Strings.isNullOrEmpty(format.getProgress())) {
//                    newestStatus = DownloadingStatus.CANCELED;
//                    ToastUtils.showToast(binding.getRoot().getContext(), binding.getRoot().getContext().getString(R.string.canceled_downloading));
//                }
//            });
//                new Thread(() -> {
//                    try {
//                        DownloadUtil downloadUtil = new DownloadUtil();
//                        downloadUtil.doDownloadVideoFile(
//                                format,
//                                downloadPath,
//                                TokenPref.getInstance(binding.getRoot().getContext()).getTokenId(),
//                                progress -> {
//                                    updateProgress(progress);
//                                    format.setProgress("" + progress);
//                                    return null;
//                                },
//                                file -> {
//                                    format.setProgress(null);
//                                    format.setDownload(true);
//                                    binding.progressBar.setProgress(0);
//                                    binding.progressBar.setVisibility(View.GONE);
//                                    format.setProgress("100");
//                                    if (onMessageControlEventListener != null) {
//                                        onMessageControlEventListener.onContentUpdate(getMessage().getId(), format.getClass().getName(), format.toStringContent());
//                                    }
//                                    //handling the behavior of cancel downloading
//                                    binding.progressBar.isCanceledLoading = false;
//                                    if(newestStatus == DownloadingStatus.CANCELED)
//                                        canceledOrFailureHandle(message, format, "", newestStatus);
//                                    else
//                                        newestStatus = DownloadingStatus.SUCCESS;
//                                    return null;
//                                },
//                                errorMsg -> {
//                                    canceledOrFailureHandle(message, format, errorMsg, newestStatus = DownloadingStatus.FAILURE);
//                                    if (onMessageControlEventListener != null) {
//                                        onMessageControlEventListener.onVideoClick(getMessage());
//                                    }
//                                    return null;
//                                }
//                        );
//                        getMessage().setStatus(MessageStatus.RECEIVED);
//                    } catch (Exception e) {
//                        CELog.e("download file failed. reason = " + e.getMessage());
//                    }
//                }).start();
//        }
//    }
//
//    @Override
//    public void onDoubleClick(View v, MessageEntity message) {
//    }
//
//    @Override
//    public void onLongClick(View v, float x, float y, MessageEntity message) {
//        if (this.onMessageControlEventListener != null) {
//            this.onMessageControlEventListener.onLongClick(getMessage(), (int) x, (int) y);
//        }
//    }
//
//    private void canceledOrFailureHandle(MessageEntity message, VideoContent videoContent, String e, DownloadingStatus status) {
//        videoContent.setProgress(null);
//        videoContent.setDownload(false);
//        message.setContent(videoContent.toStringContent());
//        binding.progressBar.setProgress(0);
//        binding.progressBar.setVisibility(View.GONE);
//        if (onMessageControlEventListener != null) {
//            onMessageControlEventListener.onContentUpdate(getMessage().getId(), videoContent.getClass().getName(), videoContent.toStringContent());
//        }
//        try {
//            String downloadPath = downloadDir + message.getSendTime() + "_" + videoContent.getName();
//            File downloadFile = new File(downloadPath);
//            boolean isDelete = downloadFile.delete();
//            CELog.e(isDelete ? "刪除檔案成功" : "刪除檔案失敗");
//        } catch (Exception e1) {
//            CELog.e("刪除檔案失敗", e);
//        }
//    }
//
//}

File: app/src/main/java/tw/com/chainsea/chat/messagekit/main/viewholder/VideoMessageView.kt
package tw.com.chainsea.chat.messagekit.main.viewholder

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.drawable.Drawable
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.webkit.URLUtil
import android.widget.Toast
import androidx.core.content.res.ResourcesCompat
import androidx.viewbinding.ViewBinding
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestBuilder
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import tw.com.aile.sdk.bean.message.MessageEntity
import tw.com.chainsea.android.common.log.CELog
import tw.com.chainsea.android.common.video.VideoHelper
import tw.com.chainsea.ce.sdk.bean.msg.MessageStatus
import tw.com.chainsea.ce.sdk.bean.msg.content.VideoContent
import tw.com.chainsea.ce.sdk.event.EventMsg
import tw.com.chainsea.ce.sdk.event.MsgConstant
import tw.com.chainsea.chat.R
import tw.com.chainsea.chat.databinding.MsgkitVideoBinding
import tw.com.chainsea.chat.messagekit.main.MessageVideoDownloadProgress
import tw.com.chainsea.chat.messagekit.main.viewholder.base.MessageBubbleView
import tw.com.chainsea.chat.util.DownloadUtil
import tw.com.chainsea.chat.util.DownloadUtil.getVideoDuration
import tw.com.chainsea.chat.util.DownloadingTask
import java.io.File

class VideoMessageView(
    binding: ViewBinding
) : MessageBubbleView<VideoContent>(binding) {
    private var binding: MsgkitVideoBinding =
        MsgkitVideoBinding.inflate(LayoutInflater.from(itemView.context), null, false)

    companion object {
        private const val SMALL_PIC_SIZE: Int = 360
    }

    private val downloadTask: DownloadingTask?
        get() {
            return message?.let {
                val videoContent = it.content() as VideoContent
                DownloadUtil.getCurrentDownloadTask(videoContent.url)
            }
        }

    private var downloadStatus = DownloadingStatus.UNDEF

    private var isLoadingVideoFirstFrame: Boolean = false

    override fun getContentResId(): Int = R.layout.msgkit_video

    override fun getChildView(): View = binding.root

    override fun rightBackground(): Int = R.drawable.msgkit_trans_bg

    override fun leftBackground() = R.drawable.msgkit_trans_bg

    override fun onDoubleClick(
        v: View?,
        message: MessageEntity?
    ) {
        CELog.d("onDoubleClick")
    }

    override fun onLongClick(
        v: View?,
        x: Float,
        y: Float,
        message: MessageEntity?
    ) {
        onMessageControlEventListener?.onLongClick(message, x.toInt(), y.toInt())
    }

    override fun showName(): Boolean = !isRightMessage

    override fun bindContentView(videoContent: VideoContent) {
        binding.thumbnailRIV.setImageDrawable(null)
        isLoadingVideoFirstFrame = false
        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this)
        }
        CoroutineScope(Dispatchers.IO).launch {
            val params = binding.occupationCL.layoutParams
            val zoomSize = getZoomSize(videoContent)
            params.width = zoomSize[0]
            params.height = zoomSize[1]
            // 如果沒高度寬度 給預設
            if (params.width <= 0) params.width = 300
            if (params.height <= 0) params.height = 600

            withContext(Dispatchers.Main) {
                binding.occupationCL.layoutParams = params
                binding.occupationCL.background =
                    ResourcesCompat.getDrawable(
                        binding.root.context.resources,
                        R.drawable.file_msg_down_bg,
                        null
                    )
                binding.thumbnailRIV.visibility = View.VISIBLE
                binding.clPlay.visibility = View.VISIBLE
            }

            setMessageStatus(videoContent.progress)
            setPreviewImage(videoContent)
        }
    }

    private fun setMessageStatus(progress: String?) =
        CoroutineScope(Dispatchers.IO).launch {
            if (progress.isNullOrEmpty()) return@launch
            val pgs = progress.toInt()
            when (message.status) {
                MessageStatus.READ, MessageStatus.RECEIVED, MessageStatus.SUCCESS -> {
                    withContext(Dispatchers.Main) {
                        if (pgs >= 100) {
                            binding.progressBar.visibility = View.GONE
                            binding.clPlay.visibility = View.VISIBLE
                        } else {
                            binding.progressBar.progress = pgs
                            binding.clPlay.visibility = View.GONE
                        }
                    }
                }

                MessageStatus.SENDING -> {
                    withContext(Dispatchers.Main) {
                        if (pgs >= 100) {
                            binding.progressBar.progress = 0
                            binding.progressBar.visibility = View.GONE
                            binding.clPlay.visibility = View.VISIBLE
                        } else {
                            binding.progressBar.progress = pgs
                            binding.clPlay.visibility = View.GONE
                        }
                    }
                }

                MessageStatus.FAILED, MessageStatus.ERROR -> {
                }

                else -> {
                    // nothing
                }
            }
        }

    // 設置影片預覽圖
    private fun setPreviewImage(videoContent: VideoContent) =
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val thumbnailUrl = if (message.content.isNotEmpty()) videoContent.thumbnailUrl ?: "" else ""
                val file = getVideoFile(videoContent)
                var request: RequestBuilder<Drawable>? = null
                file?.let {
                    withContext(Dispatchers.Main) {
                        binding.playIV.setImageResource(R.drawable.play)
                        binding.clPlay.visibility = View.VISIBLE
                        val duration =
                            if (videoContent.duration == null || videoContent.duration == 0.0) {
                                getVideoDuration(it.absolutePath)
                            } else {
                                getUpdateTimeFormat((videoContent.duration * 1000).toInt())
                            }
                        binding.tvPeriod.text = duration.toString()
                        binding.tvPeriod.visibility = View.VISIBLE
                    }
                    val videoThumbnail = thumbnailUrl.ifEmpty { it.absolutePath }
                    withContext(Dispatchers.Main) {
                        request =
                            Glide
                                .with(binding.thumbnailRIV)
                                .load(videoThumbnail)
                                .apply(
                                    RequestOptions()
                                        .override(SMALL_PIC_SIZE)
                                        .fitCenter()
                                )
                    }
                } ?: run {
                    withContext(Dispatchers.Main) {
                        binding.playIV.setImageResource(R.drawable.ic_video_download)
                        binding.tvPeriod.visibility = View.GONE
                        binding.clPlay.visibility = View.VISIBLE
                    }
                    if (URLUtil.isValidUrl(videoContent.url) || videoContent.url.endsWith(".mp4")) {
                        if (thumbnailUrl.isNotEmpty()) {
                            val thumbnailWidth =
                                if (videoContent.thumbnailWidth > 0) videoContent.thumbnailWidth else 300
                            val thumbnailHeight =
                                if (videoContent.thumbnailHeight > 0) videoContent.thumbnailHeight else 600
                            request =
                                Glide
                                    .with(binding.thumbnailRIV)
                                    .load(thumbnailUrl)
                                    .apply(
                                        RequestOptions()
                                            .frame(1000)
                                            .override(thumbnailWidth, thumbnailHeight)
                                            .fitCenter()
                                    )
                        } else {
                            isLoadingVideoFirstFrame = true
                            val mediaMetadataRetriever = MediaMetadataRetriever()
                            mediaMetadataRetriever.setDataSource(
                                videoContent.url,
                                HashMap<String, String>()
                            )
                            val bmFrame =
                                mediaMetadataRetriever.getFrameAtTime(1000) // unit in microsecond
                            withContext(Dispatchers.Main) {
                                bmFrame?.let {
                                    if (!isLoadingVideoFirstFrame) return@withContext
                                    binding.thumbnailRIV.setImageBitmap(bmFrame)
                                }
                            }
                        }
                    } else {
                        request =
                            Glide
                                .with(binding.thumbnailRIV)
                                .load(if (isGreenTheme) R.drawable.image_load_error_green else R.drawable.image_load_error)
                                .apply(
                                    RequestOptions()
                                        .override(SMALL_PIC_SIZE)
                                )
                        withContext(Dispatchers.Main) {
                            binding.clPlay.visibility = View.GONE
                        }
                    }
                }

                withContext(Dispatchers.Main) {
                    request?.diskCacheStrategy(DiskCacheStrategy.ALL)?.into(
                        object : CustomTarget<Drawable>() {
                            override fun onLoadFailed(errorDrawable: Drawable?) {
                                super.onLoadFailed(errorDrawable)
                            }

                            override fun onLoadStarted(placeholder: Drawable?) {
                                super.onLoadStarted(placeholder)
                            }

                            override fun onResourceReady(
                                resource: Drawable,
                                transition: Transition<in Drawable>?
                            ) {
                                binding.thumbnailRIV.setImageDrawable(resource)
                            }

                            override fun onLoadCleared(placeholder: Drawable?) {
                            }
                        }
                    )
                }
            } catch (ignored: Exception) {
            }
        }

    private suspend fun getZoomSize(videoContent: VideoContent): IntArray =
        withContext(Dispatchers.IO) {
            return@withContext if (videoContent.width != 0 && videoContent.height != 0) {
                zoomImage(videoContent.width, videoContent.height)
            } else if (videoContent.thumbnailHeight != 0 && videoContent.thumbnailWidth != 0) {
                zoomImage(videoContent.thumbnailWidth, videoContent.thumbnailHeight)
            } else {
                val file = getVideoFile(videoContent)
                val size =
                    if (file == null) VideoHelper.size(videoContent.url) else VideoHelper.size(file.absolutePath)
                if (size[0] > 0 && size[1] > 0) {
                    zoomImage(size[0], size[1])
                } else {
                    size
                }
            }
        }

    // 判斷是否已下載
    private suspend fun getVideoFile(videoContent: VideoContent): File? =
        withContext(Dispatchers.IO) {
            val downloadPath =
                DownloadUtil.downloadFileDir + videoContent.name
            return@withContext if (videoContent.android_local_path != null && File(videoContent.android_local_path).exists()) {
                File(videoContent.android_local_path)
            } else if (File(downloadPath).exists()) {
                File(downloadPath)
            } else {
                null
            }
        }

    private suspend fun zoomImage(
        width: Int,
        height: Int
    ): IntArray =
        withContext(Dispatchers.IO) {
            return@withContext if (width > height) {
                intArrayOf(SMALL_PIC_SIZE, (height * SMALL_PIC_SIZE) / width)
            } else {
                intArrayOf((width * SMALL_PIC_SIZE) / height, SMALL_PIC_SIZE)
            }
        }

    private fun updateProgress(progress: Int) =
        CoroutineScope(Dispatchers.Main).launch {
            binding.progressBar.progress = progress
            if (binding.progressBar.progress >= 100) {
                binding.progressBar.visibility = View.GONE
                binding.clPlay.visibility = View.VISIBLE
            } else {
                binding.progressBar.visibility = View.VISIBLE
                binding.clPlay.visibility = View.GONE
            }
        }

    private fun onDownSuccess(
        videoContent: VideoContent,
        file: File
    ) = CoroutineScope(Dispatchers.IO).launch {
        if (downloadStatus == DownloadingStatus.CANCELED) {
            canceledOrFailureHandle(videoContent)
            deleteFile(DownloadUtil.downloadFileDir + videoContent.name)
        } else {
            if (file.exists()) {
                handleDownloadSuccessful(videoContent, file)
            }
        }
    }

    private fun canceledOrFailureHandle(videoContent: VideoContent) =
        CoroutineScope(Dispatchers.IO).launch {
            videoContent.progress = "0"
            videoContent.isDownload = false
            downloadStatus = DownloadingStatus.UNDEF
            message.content = videoContent.toStringContent()
            withContext(Dispatchers.Main) {
                binding.root.post {
                    binding.progressBar.progress = 0
                    binding.progressBar.visibility = View.GONE
                    binding.progressBar.isCanceledLoading = false
                    binding.playIV.setImageResource(R.drawable.ic_video_download)
                    binding.tvPeriod.visibility = View.GONE
                    binding.clPlay.visibility = View.VISIBLE
                }
            }
            deleteFile(DownloadUtil.downloadFileDir + videoContent.name)
            onMessageControlEventListener?.onContentUpdate(message.id, videoContent.javaClass.name, videoContent.toStringContent())
        }

    private fun handleDownloadSuccessful(
        videoContent: VideoContent,
        file: File
    ) = CoroutineScope(Dispatchers.IO).launch {
        withContext(Dispatchers.Main) {
            binding.progressBar.progress = 0
            binding.progressBar.visibility = View.GONE
            binding.progressBar.isCanceledLoading = false
            binding.playIV.setImageResource(R.drawable.play)
            binding.tvPeriod.text = getVideoDuration(file.absolutePath)
            binding.tvPeriod.visibility = View.VISIBLE
            val mediaScanIntent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE)
            val contentUri = Uri.fromFile(file)
            mediaScanIntent.data = contentUri
            binding.root.context.sendBroadcast(mediaScanIntent)
        }
        videoContent.isDownload = true
        videoContent.progress = "100"
        downloadStatus = DownloadingStatus.SUCCESS
        onMessageControlEventListener?.updateReplyMessageWhenVideoDownload(message.id)
        onMessageControlEventListener?.onContentUpdate(message.id, videoContent.javaClass.name, videoContent.toStringContent())
    }

    private fun deleteFile(downloadPath: String) =
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val file = File(downloadPath)
                val isDeleted = file.delete()
                if (isDeleted) {
                    downloadStatus = DownloadingStatus.FAILURE
                }
                CELog.e(if (isDeleted) "刪除檔案成功" else "刪除檔案失敗")
            } catch (e: Exception) {
                CELog.e("刪除檔案失敗", e)
            }
        }

    override fun onClick(
        v: View,
        message: MessageEntity
    ) {
        super.onClick(v, message)
        CoroutineScope(Dispatchers.IO).launch {
            val videoContent = message.content() as VideoContent
            val downloadPath = DownloadUtil.downloadFileDir + videoContent.name
            val isFileExists = File(downloadPath).exists() || (!videoContent.android_local_path.isNullOrEmpty() && File(videoContent.android_local_path).exists())
            if (isFileExists) {
                onMessageControlEventListener?.onVideoClick(message)
//                binding.progressBar.setOnFileClickListener(null)
            } else {
                binding.progressBar.setOnFileClickListener {
                    downloadTask?.let {
                        it.task.cancel()
                        downloadStatus = DownloadingStatus.CANCELED
                        canceledOrFailureHandle(videoContent)
                        Toast.makeText(binding.root.context, binding.root.context.getString(R.string.canceled_downloading), Toast.LENGTH_SHORT).show()
                    }
                }
                DownloadUtil.doDownloadVideoFile(
                    videoContent,
                    { progress ->
                        if (downloadStatus == DownloadingStatus.CANCELED) {
                            canceledOrFailureHandle(videoContent)
                        } else {
                            EventBus.getDefault().post(EventMsg(MsgConstant.MESSAGE_VIDEO_UPDATE_DOWNLOAD_PROGRESS, MessageVideoDownloadProgress(progress, message.id!!)))
                            videoContent.progress = progress.toString()
                        }
                    },
                    { downloadFile ->
                        onDownSuccess(videoContent, downloadFile)
                    },
                    { errorMessage ->
                        downloadTask?.let {
                            deleteFile(it.path)
                            canceledOrFailureHandle(videoContent)
                        }
                    }
                )
            }
        }
    }

    /**
     * 若影片長度超過 1 小時, 格式為 h:mm:ss, e.g., 影片時長1小時3分06秒, 顯示為 1:03:06
     * 若影片長度少於 1 小時, 格式為 m:ss, e.g., 影片時長3分06秒, 顯示為 3:06
     * 若影片時長為5秒，則顯示為 0:05
     */
    @SuppressLint("DefaultLocale")
    private suspend fun getUpdateTimeFormat(millisecond: Int): String =
        withContext(Dispatchers.IO) {
            // 将毫秒转换为秒
            val second = millisecond / 1000
            // 计算小时
            val hh = second / 3600
            // 计算分钟
            val mm = second % 3600 / 60
            // 计算秒
            val ss = second % 60
            // 判断时间单位的位数
            val str: String =
                if (hh != 0) { // 表示时间单位为三位
                    String.format("%02d:%02d:%02d", hh, mm, ss)
                } else {
                    String.format("%02d:%02d", mm, ss)
                }
            return@withContext str
        }

    private enum class DownloadingStatus {
        UNDEF,
        SUCCESS,
        CANCELED,
        FAILURE
    }

    init {
        getView(this.binding.root)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(eventMsg: EventMsg<Any>) {
        when (eventMsg.code) {
            MsgConstant.MESSAGE_VIDEO_UPDATE_DOWNLOAD_PROGRESS -> {
                val messageVideoDownloadProgress = eventMsg.data as MessageVideoDownloadProgress
                if (messageVideoDownloadProgress.messageId == message.id) {
                    updateProgress(messageVideoDownloadProgress.progress)
                }
            }
        }
    }

    fun unRegisterEventBus() {
        if (EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().unregister(this)
        }
    }
}
